cmd/bio-rd/main.go

```diff
@@ -0,0 +1,201 @@
+package main
+
+import (
+	"flag"
+	"os"
+	"os/signal"
+	"syscall"
+	"time"
+
+	"github.com/bio-routing/bio-rd/cmd/bio-rd/config"
+	bgpserver "github.com/bio-routing/bio-rd/protocols/bgp/server"
+	"github.com/bio-routing/bio-rd/routingtable"
+	"github.com/bio-routing/bio-rd/routingtable/vrf"
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+)
+
+var (
+	configFilePath = flag.String("config.file", "bio-rd.yml", "bio-rd config file")
+	sigHUP         = make(chan os.Signal)
+	vrfReg         = vrf.NewVRFRegistry()
+	bgpSrv         bgpserver.BGPServer
+	runCfg         *config.Config
+)
+
+func main() {
+	flag.Parse()
+
+	startCfg, err := config.GetConfig(*configFilePath)
+	if err != nil {
+		log.Errorf("Unable to get config: %v", err)
+		os.Exit(1)
+	}
+
+	bgpSrv = bgpserver.NewBGPServer(
+		startCfg.RoutingOptions.RouterIDUint32,
+		[]string{
+			":179",
+		},
+	)
+
+	err = bgpSrv.Start()
+	if err != nil {
+		log.Fatalf("Unable to start BGP server: %v", err)
+		os.Exit(1)
+	}
+
+	vrfReg.CreateVRFIfNotExists("master", 0)
+
+	go configReloader()
+	sigHUP <- syscall.SIGHUP
+	installSignalHandler()
+
+	select {}
+}
+
+func installSignalHandler() {
+	signal.Notify(sigHUP, syscall.SIGHUP)
+}
+
+func configReloader() {
+	for {
+		<-sigHUP
+		log.Infof("Reloading configuration")
+		newCfg, err := config.GetConfig(*configFilePath)
+		if err != nil {
+			log.Errorf("Failed to get config: %v", err)
+			continue
+		}
+
+		err = loadConfig(newCfg)
+		if err != nil {
+			log.Errorf("Unable to load config: %v", err)
+			continue
+		}
+
+		log.Infof("Configuration reloaded")
+	}
+}
+
+func loadConfig(cfg *config.Config) error {
+
+	for _, ri := range cfg.RoutingInstances {
+		err := configureRoutingInstance(ri)
+		_ = err
+
+	}
+
+	if cfg.Protocols != nil {
+		if cfg.Protocols.BGP != nil {
+			err := configureProtocolsBGP(cfg.Protocols.BGP)
+			if err != nil {
+				return errors.Wrap(err, "Unable to configure BGP")
+			}
+		}
+	}
+
+	return nil
+}
+
+func configureProtocolsBGP(bgp *config.BGP) error {
+	// Tear down peers that are to be removed
+	for _, p := range bgpSrv.GetPeers() {
+		found := false
+		for _, g := range bgp.Groups {
+			for _, n := range g.Neighbors {
+				if n.PeerAddressIP == p {
+					found = true
+					break
+				}
+			}
+		}
+
+		if !found {
+			bgpSrv.DisposePeer(p)
+		}
+	}
+
+	// Tear down peers that need new sessions as they changed too significantly
+	for _, g := range bgp.Groups {
+		for _, n := range g.Neighbors {
+			newCfg := BGPPeerConfig(n, vrfReg.GetVRFByRD(0))
+			oldCfg := bgpSrv.GetPeerConfig(n.PeerAddressIP)
+			if oldCfg == nil {
+				continue
+			}
+
+			if !oldCfg.NeedsRestart(newCfg) {
+				bgpSrv.ReplaceImportFilterChain(n.PeerAddressIP, newCfg.IPv4.ImportFilterChain)
+				bgpSrv.ReplaceExportFilterChain(n.PeerAddressIP, newCfg.IPv4.ExportFilterChain)
+				continue
+			}
+
+			bgpSrv.DisposePeer(oldCfg.PeerAddress)
+		}
+	}
+
+	// Turn up all sessions that are missing
```

Comment:
Thats all missing sessions and the sessions teared down above due to NeedsRestart?

Reply:
Correct.